### DISCUSSION TITLE
Editing dotfiles and some other questions

### DISCUSSION THREAD

**Discussion by HunterretnuH** [nwg-piotr/nwg-shell#37] [Q&A]
Hello

First of all thank you for all the work you've put into this project. It's awesome that finally swaywm may get consistent basic GUI and configuration tools without having to glue together lots of different sometimes not fitting well parts yourself. I hope for this project to be a good base for further personal customization.

After testing it a little bit in VM I have some doubts so I decided to ask here for clarification.

On the project page such information can be found "We discourage from editing .dotfiles manually. There’s a GUI for that. The configuration files are in json format, which does not forgive mistakes." and in the sway config itself there is such information "# The files we include below will be created / overwritten by nwg-shell tools".

I may be wrong but I believe that most of users who could possibly use nwg-shell will expect a way to modify dotfiles.
Unless nwg-shell is so feature rich that it can handle absolutely any sway configuration from GUI there will always be a need to edit dotfiles manually

But aside from above mentioned statements regarding dot file modifications there actually is currently a separation between settings that can be modified and the ones that can't and there is also option to disable specific settings it looks like it's prepared to allow dot file modification by user.

So here are my questions:
1. Are the settings in ./config/sway/config just defaults which can be safely modified without breaking anything in GUI (of course aside from settings set in few files included at the start)? 
2. Do you plan to keep this separation between settings that can be modified, the ones that can't and allow disabling every part of GUI configuration?
3. How to backup settings set in GUI? Normally I have my dotfiles backed up on github and with the help of some scripts I could easily set up my environment on fresh linux installation. But for the settings set in GUI I can't just copy dotfiles as they won't be loaded by GUI later and they will be overwritten. There was a mention of JSON files so I assume that's were all settings are kept. Is it possible to just back up these JSON files and just copy them after new running nwg-shell-installer? How would you approach this and are there any problems to be expected?
3. Is it stable and can be used right now without introducing any big problems/unstability into the system?
4. In the current project state is it enough to update packages after installation and everything should work fine or are there still big changes introduced which may require reinstallation of whole nwg-shell or worse clean linux installation and installation of nwg-shell again?
5.  Just curious. Are you the only person working on it?

**Comment by nwg-piotr** [comment]
Hi,

Not bad set of questions. :rofl:

If it comes to editing config files: I mean mostly the shell-specific json files. All the content is accessible through GUI, so I don't support people who spoil their config by manual edition. Every single space matters in the json format.

Feel free to edit your main sway config file, provided that you know what you're doing. Editing what we include into it is of course possible as well, but your changes will be overwritten right after you press the "Apply" button in nwg-shell-config, so it makes little sense.

> How to backup settings set in GUI?

Basically it's all stored in `~/.config/nwg-*/` and `~/.config/sway` directories, but some data are also placed in the `~/.local/share/` dir. I've been thinking about a backup/restore utility, but nobody asked for it so far. 

> Is it stable

Stable enough, as far as I know. Please report bugs, if you find any.

> is it enough to update packages after installation and everything should work fine

Mostly. Sometimes there's a need to modify existing config files, but I do my best to avoid such situations. If it happens, you'll see a notification with the "Update now" / "Later" buttons. Also the "Updates" button in the config utility will have a dot:

![image](https://user-images.githubusercontent.com/20579136/202317615-280657f9-8fb1-47c6-9c83-42c2e1b5f126.png)

Some minor changes to default configs are being introduced without adding them to user config files. They only affect fresh installations, or use of the `nwg-shell-installer` command. Possibly the future backup/restore utility could have an option to restore defaults selectively.

> Are you the only person working on it?

On the shell as a whole - yes, at least so far. Certain components have their co-authors, and their contribution cannot be overestimated.

**Comment by HunterretnuH** [comment]
Thank you for the explanation. 
I still have some questions and just finished writing huge block of text with questions and some suggestions, but I decided to postpone posting it as I just realized at the end that the answer for one question may allow me to remove huge part of it and not waste your time.

"Feel free to edit your main sway config file, provided that you know what you're doing. Editing what we include into it is of course possible as well, but your changes will be overwritten right after you press the "Apply" button in nwg-shell-config, so it makes little sense."

And that's what I'm afraid off, but maybe unnecessarily. By "Editing what we include into it" do you mean explicitly "include" keyword or are there any settings in main config file that may be overwritten by clicking "Apply" button?

I can see at least one settings which is related to GUI - "exec ~/.azotebg". Although this line itself probably isn't modified by GUI it will make wallpaper configuration in GUI not work if it is chenged. Shouldn't it be moved to "autostart" and option to disable wallpapers in GUI should be added? 
Also why there is a need to call separate bash script to set background, why not just move it to config files? I can see line "killall swaybg 2>/dev/null" so maybe that's the reason,but I don't remember having any problem with multiple instances of swaybg being started after reloading sway settings.

**Reply by nwg-piotr** [reply]
> And that's what I'm afraid off, but maybe unnecessarily. By "Editing what we include into it" do you mean explicitly "include" keyword or are there any settings in main config file that may be overwritten by clicking "Apply" button?

```text
include variables
include outputs
include autostart
include workspaces
include keyboard
include pointer
include touchpad
```

I mean the files named "variables", "includes", etc. They execute as if they were a part of the sway config file, but their content is created by GUI. If you uncheck the appropriate "Use these settings" check button, the file will still exist, but empty. That's why there's no need to delete e.g. `include keyboard` from the sway config file. Besides these files, you can include whatever else you want.

> I can see at least one settings which is related to GUI - "exec ~/.azotebg". Although this line itself probably isn't modified by GUI it will make wallpaper configuration in GUI not work if it is chenged. Shouldn't it be moved to "autostart" and option to disable wallpapers in GUI should be added?

I'm not sure what you mean. The preinstalled `~/.azotebg` file points to the default bitmap file in the backgrounds folder. If you make any change in Azote, it will overwrite `~/.azotebg`. If you don't want to use Azote as your wallpaper manager, just remove `exec ~/.azotebg` from your sway config file.

> Also why there is a need to call separate bash script to set background, why not just move it to config files?

Because the script content is managed by Azote, whose users might have never heard about nwg-shell, sway or even Wayland.

**Comment by HunterretnuH** [comment]
> I'm not sure what you mean. The preinstalled ~/.azotebg file points to the default bitmap file in the backgrounds folder. If you make any change in Azote, it will overwrite ~/.azotebg. If you don't want to use Azote as your wallpaper manager, just remove exec ~/.azotebg from your sway config file.

I just meant that I feel like there shouldn't be any lines which modifications could impact GUI configuration in main sway config file.
If user deletes line "exec ~/azotebg" the wallpaper button will stop affecting background. So change in main config will "break" GUI.
Line "exec ~/.azotebg" could me moved to "include autostart" file and there could be a checkbox in the GUI to disable background configuration in GUI (like with other things already) which could grey out "wallpaper" button.
This one is just a matter of how GUI config and Dotfile config is separated and being consistent about it throughout the project.
It's not a problem for me and I just wanted to know why it's handled like that.

I will postpone posting any of my suggestions until I use nwg-shell more and understand it better. Just to be fully sure that I understood you correctly, here is my understanding of how updates and "Apply" button currently work:
1. Updates:
	* updates itself does not modify sway configs itself in any way (aside from maybe some fixes in included files);
	* there are no changes to default configs. Such changes will only affect fresh installation;
	* updates that require manual changes by user are marked by icon with dot;
2. Modifying settings in GUI and  clicking "Apply" button will only modify content of :
``` include variables
include outputs
include autostart
include workspaces
include keyboard
include pointer
include touchpad
```
end absolutely nothing below these lines in main config file.

Please just confirm if I understood you and correct me if I'm wrong.

Z góry dziękuję i przy okazji pozdrawiam rodaka :)

**Reply by nwg-piotr** [reply]
> Line "exec ~/.azotebg" could me moved to "include autostart" file and there could be a checkbox in the GUI to disable background configuration in GUI (like with other things already) which could grey out "wallpaper" button.

I see no such need. Azote is much older than nwg-shell, and takes care of the background configuration. No need to reinvent the wheel.

> absolutely nothing below these lines in main config file

The config utility doesn't modify the main config file at all. Just includes.



> Z góry dziękuję i przy okazji pozdrawiam rodaka :)

O, jak miło. :) Użytkowników z Polski można policzyć na palcach jednej ręki.
